### DISCUSSION TITLE
Berry examples repository

### DISCUSSION THREAD

**Discussion by darrylb123** [arendst/Tasmota#17490] [General]
I've just spent a while writing some berry code. I had never used Berry before and, it seems, neither has most.  There is very little Berry code to be found.
The documentation in Tasmota and Berry is very complete, except there are very few examples.
The Tasmota Cookbook seems aimed at showing off what Berry can do, very admirable but not helpful to a beginner.
Can I suggest that an area be made available that example code may be submitted to help newbies like myself.

I'd be happy to submit a few.

**Comment by sfromis** [comment]
https://github.com/tasmota/Berry_playground was created recently, with the intention of @dkebler getting it ready around Yuletide.

**Comment by dkebler** [comment]
best laid plans.... intention now is to get to this next week.  @darrylb123 if you would like to assist help is appreciated.  There will be a way to submit a pr to share your code efforts.

**Reply by darrylb123** [reply]
@kebler How is progress with the repo

**Reply by darrylb123** [reply]
@dkebler How is progress with the Berry playground?

**Comment by darrylb123** [comment]
Happy to.
I've not tried to do anything substantial but have done a watering system
with flowmeters in berry.
I have implemented:
Timers
Cron
Read and write counters
Read and write tasmota relays
Mqtt publish
Mqtt subscribe
Json handling in and out

So I could produce examples of those very easily.

**Reply by LucidEye** [reply]
Hello, I am trying to figure out how to setup a simple switching logic script.  I am using an ESP32 and a 4 channel relay board to control 4 zone irrigation valves for watering my garden.  I also have a 5th relay which I need to turn on only when relays 1-4 are on, and likewise make sure relay 5 is off whenever relays 1-4 are off.  This 5th relay applies power to the 24 VAC transformer which energizes the valve solenoids. Can you please provide an example of how I might accomplish this using the Berry script? I have looked extensively around the internet and I can not find very many useful or relevant examples at all in this obscure scripting language.  I am not a programmer so this is language is really abstract and confusing to me.  I wouldn't even know how to begin coding something like this.

The basic logic of what I need to happen is something like this:
IF Power1-4 = ON THEN Power 5 = ON ELSE Power1-4 = OFF THEN Power5 = OFF

That's it... I just don't have any clue as to the proper syntax as I am not familiar with this at all.
It would be nice if I could just lump Power1 through Power4 in one IF-THEN-ELSE command, but I realize I may need to make a separate IF-THEN-ELSE function for each relay.... but I still have no clue how to do that in this language.

Any help would be appreciated.  Thank you.

**Reply by sfromis** [reply]
Something like this: (not tested a lot, but appears to work)
```
class MasterPower
  def init()
    tasmota.add_driver(self)
  end
  def set_power_handler(cmd, idx)
    var pow = tasmota.get_power()
    var anypower = pow[0] || pow[1] || pow[2] || pow[3]
    if anypower != pow[4]
      tasmota.set_timer(0, /-> tasmota.set_power(4, anypower))
    end
  end
end
MasterPower()
```

**Reply by LucidEye** [reply]
Thank you!  However, could you please include some basic comments in the code to explain what each line is actually doing?  Without some kind of "translation" this is all still abstract and meaningless to me.
This is what I think I understand so far...

class MasterPower     #Defines MasterPower as the object name for this whole loop and all functions contained within?
  def init()     #???
    tasmota.add_driver(self)     #Allows this function to respond to messages from Tasmota and be used as a driver for hardware?
  end
  def set_power_handler(cmd, idx)    #Allows following functions to read/write power-state inputs and outputs?
    var pow = tasmota.get_power()     #Defines variable "pow" as being equal to the output state of the command "tasmota.get_power()"
    var anypower = pow[0] || pow[1] || pow[2] || pow[3]     #Defines realys 1-4's states as being lumped under the "anypower" variable 
    if anypower != pow[4]     #If any relay 1-4 is "true" then relay 5 is also "true"???
      tasmota.set_timer(0, /-> tasmota.set_power(4, anypower))     #If relay 5 is "true" then use these Tasmota settings for the 5th relay
    end
  end
end
MasterPower()

By the way, this script didn't work.....  I get the following error whenever I activate any relay 1-4:
![Screenshot at 2023-08-11 12-05-25](https://github.com/arendst/Tasmota/assets/12551280/b2a5288c-ddb6-4809-91ff-9a4100faab3c)
Although I don't know if I'm entering the code correctly.... I just copied and pasted this into the Berry console and clicked Run.  
Should I have saved it as a file and uploaded it to the file system as a .be or .bec file?
Also, how do I clear any code I have already entered into the console so I can restart from scratch with a new and edited script?
Again, thank you for your help and your patience ... I am starting to slightly grasp this after cross-referencing your script with all the documentation... Just need a bit more help understanding.

**Reply by sfromis** [reply]
Well, you need 5 power channels for it to work. You talked about first 4 channels as the valves, and then a 5th for the transformer. That's why it fails for you. If you only have 3 valves instead of 4, and the 4th channel is for the transformer, you can remove the `|| pow[3]` part and replace the 4 by 3 where used.

Your comments about the code are largely correct, but still a couple of details....
1. First line defines a class, not an object, and last line instantiates the class to create an object.
2. The function of `init()` is initialization code for when the class is instantiated (creating an object)
3. The `tasmota.add_driver` allows the class instance to [receive callbacks](https://tasmota.github.io/docs/Berry/#creating-a-tasmota-driver) from Tasmota in many different situations
4. `set_power_handler()` is just one of those, and it is only to be notified when a power status change happens.
5. `tasmota.get_power()` accesses a list of power states for all channels, boolean values
6. Index numbers in Berry starts with 0, and with 1 in Tasmota outside of Berry. Hence `pow(2)` matches `Power3` etc
7. `tasmota.set_power()` simply changes the state of one power channel
8. Tricky thing is using that `tasmota.set_timer(0, ...`. The reason is that the `set_power_handler` callback is not allowed to change any states, therefore the set_power call is deferred for 0 milliseconds, meaning after exit from set_power_handler

Berry code entered in the Berry REPL console is not saved anywhere. You can create a file `autoexec.be` with Berry code to be loaded after boot. When you use the console, the new code will in most cases supersede the old, but when using `tasmota.add_driver` the old iteration will persist. You can either reboot, build Tasmota with a development option providing a Berry restart button in the Berry REPL console, or you can change the `init()` method to:
```
  def init()
    tasmota.remove_driver(global.MasterPower_instance)
    global.MasterPower_instance = self
    tasmota.add_driver(global.MasterPower_instance)
  end
```
This allows the driver to remove the old version of itself, and then register the new instance.

**Reply by LucidEye** [reply]
I take it back... the script does indeed work.
I did not have my test hardware configured correctly.
My test hardware is only using a 4-channel relay (no 5th relay) so I needed to edit the script to this:
class MasterPower
  def init()
    tasmota.add_driver(self)
  end
  def set_power_handler(cmd, idx)
    var pow = tasmota.get_power()
    var anypower = pow[0] || pow[1] || pow[2]
    if anypower != pow[3]
      tasmota.set_timer(0, /-> tasmota.set_power(3, anypower))
    end
  end
end
MasterPower()

Now whenever relays 1-3 are on/off, relay 4 is also on/off!
I also noticed that if try to manually or remotely activate relay 4 when relays 1-3 are not on, it automatically gets shut off instantly, which is actually great, it will prevent the 24 VAC transformer from accidentally getting turned on when it shouldn't be.
I guess I make it permanent across reboots by uploading the script to the file system as a .be file?

**Reply by sfromis** [reply]
Yes, as mentioned, you can create the file `autoexec.be`. You can either have the code there, or load it using something like
`load('MasterPower.be')`

**Reply by LucidEye** [reply]
Thank you again very much for explaining all of this in clear detail.  
I have it working as desired on my test hardware right now.   
I will watch it for a few days before I try running it on the actual irrigation system.
Currently I have my 3rd-party automation software (Mycodo) managing this switching logic over MQTT.
But it takes 8 discrete automation functions (2 for each relay) to accomplish what that simple berry script can do.
![Screenshot at 2023-08-11 13-47-02](https://github.com/arendst/Tasmota/assets/12551280/efdb70a7-dfa1-42ea-af8b-814f251af897)
And once in a great while, the ESP32 will fail to receive an mqtt payload due to WiFi interference, and the relays don't turn on (or off) in unison.  I much prefer the micro-controller to be handling this switching logic, as it should be.... much more reliable.
If it is OK with you, I'd like to post this code with all of the corrected code comments in the Berry code section... seems like there isn't a whole lot in there yet, and I know this will undoubtedly be useful for others.
Thank you again.

**Reply by LucidEye** [reply]
2 more questions...
Once I have the autoexec.be file uploaded to the file system, I can delete the .be file once I see the .bec file appear, correct?
Once the autoexec.bec file is created, does it get backed up when I do a Backup Configuration, or do I need to upload the autoexec.be file every time I need to re-flash a new ESP32 with these configuration settings?

**Reply by sfromis** [reply]
Sure, you *can* delete the `.be` file, but usually there is no good reason to do that. Do keep a backup somewhere.

And no, the `Backup Configuration` function does not back up any files you have in the file system. So, yes, you'll have to recreate the file for new devices, unless you do a custom build including a pre-populated file system.

And I put no restrictions of what you do with the sample code I posted here :smile:

**Comment by btsimonh** [comment]
this would really help me....
simple things like how to trigger a berry function from a digital input/hadrware button evade me after 2-3 hours of playtime.

All I want to do is to trigger something from MQTT, set a 'relay', then count pulses from a 'button', and turn off the relay after a configured number. (auto fish feeder - pulses -> portions).

@darrylb123 - I found your repo here 
https://github.com/darrylb123/tasmota/blob/main/flowmeter.be
and it has most of what I need apart from reading a button, which I think I can do from a rule (but there is a better way?). :).  thankyou.

**Reply by barbudor** [reply]
If the GPIO is defined as a Button or a Switch then all Tasmota rules trigger can be applied, once you have eventually detached such button/switch from relay if there is one

Alternatively you can use as a 'gpio' (define nothing or INPUT in the template/module configuration) and poll that in a timer based function 'a la Arduino'

**Reply by btsimonh** [reply]
Thankyou @barbudor - I went the rule trigger route...

with the example from @darrylb123 repo, I have come up with the below (in about an hour :) ).

Things to highlight:

Repeatedly running code in the berry console is cumulative - especially and obviously adding tas rules :).
At least when developing, best to remove then add.... else you end up restarting a lot.

It's not obvious that the function added with add_cmd should end with a tas response to be sent (e.g. `tasmota.resp_cmnd_done()` ).
Else I got a cryptic error `BRY: ERROR, bad json: OFF`

br,
Simon

```
# A simple Berry example.
# Operates an automatic pet feeder (an old CAT feeder)
# the purpose here is to respond to a commmand 'feed n'
# if n > 0, then a motor is started, and a counter set to n
# every time switch1 transitions to ON, a counter is decremented.
# when the counter reaches zero, the motor is turned off.
# in my case, I've used a spare hbridge to provide the motor drive, so setup relay1, relay2 as the hbridge controls.
# I've set switch1 as the switch operated by the feeder deliveing one portion of food.
# note the setting of swicth options from berry.
# note the REMOVAL of rules and cmnd before adding.

import json
import string
#import mqtt
#import persist

# define relays for motor hbridge
var forward = 0
var backward = 1

# a variable to count how many portions we have fed this time
var portion_counter = 0

def feed(val)
  if val
    # turn motor forwards if val != 0
    tasmota.set_power(forward,true)
    tasmota.set_power(backward,false)
    portion_counter = val
    print("feed", val)
  else
    # turn off motor
    tasmota.set_power(forward,false)
    tasmota.set_power(backward,false)
  end
end

# this is called by using 'feed' in tas.
# note that it needs to respond... else you get a cryptic 'BRY: ERROR, bad json: OFF'
def feed_cmnd(cmd, idx, payload)
  print("cmnd from tas:",cmd, idx, payload)
  feed(int(payload))
  tasmota.resp_cmnd_done()
end

# example of direct mqtt - we will just use a cmnd
# hence not tested.
#def feed_message(topic, idx, payload_s, payload_b)
#	var feed_msg = json.load(payload_s)
#  
# 	if feed_msg['state'] == "start"
#    feed(4)
#  else
#  end
#end
#def subscribes()
#  mqtt.subscribe("cmnd/feed/control",water_message)
#end
#tasmota.remove_rule("MQTT#Connected=1")
#tasmota.add_rule("MQTT#Connected=1", subscribes)


# this will be called from a rule in tas
def buttonTriggered()
  portion_counter = portion_counter - 1
  print("Switch activated")

  if portion_counter <= 0
    feed(0)
  end
end

# set switch 1 to report state, not toggle
tasmota.cmd("SwitchMode1 1");
# disconnect switches from relays
tasmota.cmd("SetOption114 1");

# note:
# if using the tas berry console, best to remove the rules first.
# because the old ones remain, and berry gets called multiple times.
# call our function every time the switch report on
tasmota.remove_rule("Switch1#State=1")
tasmota.add_rule("Switch1#State=1", buttonTriggered)

# our feed command - use like "feed 4" to feed a 4 portion meal.
tasmota.remove_cmd('feed')
tasmota.add_cmd('feed', / cmd, idx, payload -> feed_cmnd(cmd, idx, payload))
```

**Comment by darrylb123** [comment]
I had the need to do something similar on another application.
It was interesting as I found that a closure was the way to achieve the
result. Until that point, I couldn't understand what a closure was used for.
Given that I had to maintain state within a time delay loop.
https://github.com/darrylb123/wavetank/blob/main/softstart.be

**Reply by btsimonh** [reply]
I found another interesting construct using global.xxx where xxx is a variable you declare in the script.
When developing a 'driver' I found it a pain to have to keep restarting tas for it to remove the last iteration I had pasted in the berry console.
I could not get exception handling to work, but see https://github.com/tasmota/Berry_playground/issues/2 for my workaround...

**Reply by btsimonh** [reply]
maybe place your closure code in an issue there - again it's innovative real-world use of berry, not at all obvious from any docs I've read!

**Comment by darrylb123** [comment]
Is that what we're supposed to do? Submit it as an issue.
I've been expecting to add files to the repository once it was ready for
use?

**Reply by btsimonh** [reply]
no - I'm just playing fast and loose :).  I have a million projects on the go at a time, and will have forgotten all my new berry knowledge by next week - so just needed somewhere where I could stash it where it may be useful to others.
I've suggested they enable the wiki on the repo - that's a good way to get the public to contribute to what is effectively docs....  but can be a little uncontrolled.

**Reply by sfromis** [reply]
As things are now, it looks like @dkebler dropped the ball (?) when it comes to getting https://github.com/tasmota/Berry_playground active.

**Comment by darrylb123** [comment]
Yes, I've tried to prod him a couple of times, with no response.
I have a little time, I'd be happy to assist getting it up and running.

I assumed that it would be treated like a code repository, with some kind
of structure and examples by others would be in the form of pull requests.

A wiki might be difficult to keep some structure.

**Comment by dkebler** [comment]
Yes I've dropped the ball but it's still dribbling along the floor.  Actually in the last month I have coded up (in node) a dev environment so one can work on berry code and have it automatically fetched by a tasmota device on changes and then have the device reboot.  At least for me it's really speed up my dev cycle.  I'd like to share it along with some apps (e.g. weather station, thermostat/hvac) and utility modules I've coded but like many not alot of extra time to document it and put it at playground (or linked).      

what I think would be helpful at least it to activate the "discussion" of the playground repo so at least those interested can easier communicate and have a record of such.   I don't have permission to do that.   @sfromis  or others would have to do it.  Also anyone else who want publish rights to the repo should just ask  @sfromis as my heart is bigger than my time when it comes to making the playground a thing.

BTW a good berry tool I found was this.  https://github.com/fmtr/tappack.   So there is reason for the playground if only there was time for it.

**Reply by sfromis** [reply]
I'm not the one to ask about permissions; I think that Jason2866 has the powers.

**Comment by darrylb123** [comment]
Opening the discussion sounds great. There should be some agreement on the
structure and content.
It would be a shame for it just to end up as a code dump.

I'd like to see small concise files demonstrating a concept or feature. The
structure might be based on the reference manual, with an example for each
concept in each file, so a chapter might end a with several files in the
playground.

**Reply by dkebler** [reply]
I agree with what you say.  Maybe the playground repo can also hold a "utility" modules folder.  But first let's see if we can get @sfromis attention to add the discussion to playground (we can try a discord if need be).  At the very least why don't you close this issue and we can use the playground issues for now.    Here is link to a playground issue to continue.  

https://github.com/tasmota/Berry_playground/issues/4#issue-1714025542

**Reply by dkebler** [reply]
Now there somewhat of a structure and instructions for making contributions to the playground along with a discussion system so maybe @darrylb123 we can continue there.

https://github.com/tasmota/Berry_playground

https://github.com/tasmota/Berry_playground/discussions/3

**Comment by rug1024** [comment]
First of all: I am not a programmer, but trying to learn and understand. So you may laugh at me or ignore this posting, but I have not been able to find help elsewhere. Read Berry Cookbook, studied [other scripts](http://sfromis.strangled.net/tasmota/berry/github-repositories) , etc. , even asked AI (CoPilot), but those results are not working. By the way: CoPilot can create complex working Tasmota-Rules, it just fails with Berry.

My Hardware: ESP32, MAX7219 LED Matrix Display, Brightness sensor BH1750.
What I am trying to do (in Berry-Script): set the brightness (DisplayDimmer) of the LED-Display according to the surrounding brightness (4 steps), executed every 30 seconds.
Working Berry-Commands I have found out so far: **tasmota.read_sensors("BH1750")** (output needs formatting) and **tasmota.cmd("displaydimmer 10")**

I currently have this task working with Tasmota-Rules, but I also want it to work in Berry, mainly for learning purposes so I don't have to bother you with stupid questions in the future. Really looking forward to your hints. Thanks in advance! Regards...

**Comment by Jason2866** [comment]
@rug1024 Questions are welcome. Better to open a new discussion with a specific headline.
Way better chance to get an answer than posting the question in a not really related discussions topic.

**Comment by dkebler** [comment]
@rug1024   even better would be to open a discussion topic at the berry playground.  Then you can get Berry specific help.

https://github.com/tasmota/Berry_playground/discussions/categories/get-help-q-a

also for "live" help you might try discord.  THere is a berry channel in the Tasmota server.  But you should really have a specific question.  For general help the link above is better.

**Comment by sfromis** [comment]
To see a lot of sample Berry code, you can use this long list of Github repositories I've collected; the common thing is that they include Berry code, in almost all cases related to Tasmota.
http://sfromis.strangled.net/tasmota/berry/github-repositories

**Comment by chaertl1** [comment]
Does anybody know the proper way to define constants and enums in Berry Script ?

**Reply by barbudor** [reply]
There's no such concept in Berry
One of the closet thing you can do is create a class with static var, using the old convention of Upper case to denote constant.
```
class MyConst
  static var A = "a" 
  static var B = "b"
end

var some_var = MyConst.A

If some_var == MyConst.B
  do_something()
end
```

**Reply by sfromis** [reply]
In some cases, it can be an useful trick to have such a constant include behavior. It can refer to any object, maybe a class, a class instance, or even be a lambda for a function call:
```
class MyConst
  static var A = /->do_something()
  static var B = /->print('doing something else')
end

var some_var = MyConst.A
some_var()
```
Of course, it still works to use such constants in the "classic" way with simple assignment/compare.

**Reply by chaertl1** [reply]
OK this works, but can that static vars be protected from being changed ?

**Reply by s-hadinger** [reply]
It depends of "protecting against what or whom"? There is always a way in Berry to supersede any definition (it's a feature).

So generally speaking:
- protecting from being changed: no
- unless you solidify (i.e. burn in Flash), but even in this case, there is always a way for an evil actor to hide the value and supersede with another class that has different values

**Reply by sfromis** [reply]
No. Just do not make bad code messing with it. The prefixed class name does not make it likely to be "just a typo". If you have the variable content be a class instance, that could of course also hold mutable data.

The Berry language does not have the concepts of defining variable as private to a specific scope, or final to disallow modification.

**Comment by chaertl1** [comment]
Would it be a solution to use functions returning the desired constant value to prevent the constants to be changed?

**Reply by s-hadinger** [reply]
I repeat myself: what kind of risk do you fear?

If you use a function, nothing is preventing the bad user to define a new function returning another value

**Reply by sfromis** [reply]
When you create a function, you are merely creating an object which is stored in a variable named after the function. And any variable content can be replaced.

Why is it important to you to prevent change of constant values? Premature optimization?

**Reply by chaertl1** [reply]
It is easy to write an = instead of == in a condition that changes the variable instead of comparing it.

**Reply by barbudor** [reply]
The good thing is that Berry is not C and this won't  pass compilation  in a test
```
if some_var = another_var
```
Is illegal

**Reply by sfromis** [reply]
The highest risk of accidental overwrite may be having two errors at the same time, like to write:
`if MyConst.A == some_var MyConst.B = some_var`
while the intention was:
`if MyConst.A == some_var || MyConst.B == some_var`

If you **want** to do an assignment in a single condition, that would be the walrus:
`if some_var := another_var`
However, the `if` would not work to do a compare.
This is the only way of doing assignment inside a syntax usually doing other things.
